### PR TITLE
Support anyOf in JSON schema (#2081)

### LINKED ISSUE #2081: [FEATURE] JsonSchema - anyOf
**Is your feature request related to a problem? Please describe.**
I am using JsonSchema and cannot find a way to use the anyOf specification. It would be great if langchain4j supported it since openai does. 

- https://platform.openai.com/docs/guides/structured-outputs#supported-types
- https://platform.openai.com/docs/guides/structured-outputs#for-anyof-the-nested-schemas-must-each-be-a-valid-json-schema-per-this-subset

**Describe the solution you'd like**
Add another implementation of `JsonSchemaElement` e.g. `JsonAnyOfSchema`.

### PR DESCRIPTION
**Pull request by mareksvatos**
## Issue

Closes https://github.com/langchain4j/langchain4j/issues/2081

## Change

I have created a pull request in `dev.ai4j.openai4j` which adds `anyOf` support to the openai client https://github.com/ai-for-java/openai4j/pull/45. This change needs to be merged before pull request in this repo.

I have added `JsonAnyOfSchema` as a new implementation of `JsonSchemaElement` to support `anyOf` spec for json schema while using structured outputs. 

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module

- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonObjectSchema.java [hunk 1]
```diff
@@ -241,6 +246,30 @@ public Builder addEnumProperty(String name, List<String> enumValues, String desc
...
         }
 
+        /**
+         * Adds a single anyOf property to the properties of this JSON object.
+         * Please note that {@link #required(List)} should be set explicitly if you want the properties to be mandatory.
+         * @see #addAnyOfProperty(String, List, String)
+         * @see #addProperty(String, JsonSchemaElement)
+         * @see #properties(Map)
+         */
+        public Builder addAnyOfProperty(String name, List<JsonSchemaElement> anyOf) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonObjectSchema.java]: We do not have convenience methods for complex types (objects, arrays), so please remove these new ones as well.
I am not sure they will be widely used, but they pollute the API quite a lot. We can always add them in the future if needed.

File: langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonAnyOfSchema.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(description, anyOf);
+    }
+
+    @Override
+    public String toString() {
+        return "JsonAnyOfSchema{" +
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonAnyOfSchema.java]:
```suggestion
        return "JsonAnyOfSchema {" +
```
nit

File: langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonAnyOfSchema.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                "description='" + description + '\'' +
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonAnyOfSchema.java]:
```suggestion
                "description = " + quoted(description) +
```
nit

File: langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonAnyOfSchema.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                ", anyOf=" + anyOf +
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonAnyOfSchema.java]:
```suggestion
                ", anyOf = " + anyOf +
```
nit

File: langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonAnyOfSchema.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                '}';
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonAnyOfSchema.java]:
```suggestion
                " }";
```
nit

**Comment by langchain4j** [comment]
@MarekSvatos thank you!
I have merged https://github.com/ai-for-java/openai4j/pull/45 and released `openai4j` 0.24.0

**Review by ghost — COMMENTED** [review]
@MarekSvatos, thanks a lot, good job!

It would be fantastic to also support this on the AI Service level. Any ideas how to make it work?

File: langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonObjectSchema.java (hunk 1 above)
Comment (mareksvatos) [reply langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/json/JsonObjectSchema.java]: Removed

**Comment by mareksvatos** [comment]
> @MarekSvatos thank you! I have merged [ai-for-java/openai4j#45](https://github.com/ai-for-java/openai4j/pull/45) and released `openai4j` 0.24.0

I have updated the version in this pull request. https://github.com/langchain4j/langchain4j/pull/2148/commits/7ab89c05b4e4ac1f025331c631b6f1654ec09e9f

**Comment by mareksvatos** [comment]
> @MarekSvatos, thanks a lot, good job!
> 
> It would be fantastic to also support this on the AI Service level. Any ideas how to make it work?

I think it could be solved by using Jackson. Applying `@JsonTypeInfo` + `@JsonSubTypes` annotations is a common way to deal with polymorphic serialization/deserialization. Example is included in my test.

User would have to 

1. Create an interface and its subtypes
2. Annotate the interface accordingly 

Langchain4j would have to

1. When interface is part of the pojo(as a field or an array) scan for its `@JsonSubTypes` annotation and collect possible subtypes
2. For each subtype create corresponding `JsonSchema` and add them all to `JsonAnyOfSchema`
3. Deserialization should work out of the box

**Review by ghost — APPROVED** [review]
@MarekSvatos thanks a lot!

**Pull request merged**